### PR TITLE
make FieldValueTypeInformation creators take a TypeDescriptor parameter

### PR DESCRIPTION
**Pull request by tilgalas**
This is a 2nd PR in the series of PRs (based on the now closed #31648, with the first PR being #31785) whose ultimate goal is to add support for generic classes to schema providers. 

`FieldValueTypeInformation` creators will now accept a `TypeDescriptor` parameter describing the field's containing class, that will let them infer more accurate type information about that field. For example - consider a `MyClass<T>` class - now with a `TypeDescriptor` of `MyClass<String>` and the getter of type `T` the type resolver can infer the type of field to be `String` , see the added test class that shows the new functionality

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by tilgalas** [comment]
Run Java PreCommit

**Comment by tilgalas** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by tilgalas** [comment]
Run Java_Hadoop_IO_Direct PreCommit

**Comment by damccorm** [comment]
@damondouglas could you please take a look at this one?

**Review by damondouglas — CHANGES_REQUESTED** [review]
@tilgalas Thank you for working on this! Would you mind to first remove the `"nullness"` value in the `@SuppressWarnings` annotation for the involved classes in this PR? Please consider this suggestion as non-blocking but I would like to see if this is possible with your helpful changes. As a second phase, may we also consider removing the `"rawtypes"`? Again, non-blocking but it's something we would like to strive for to improve the code quality.

**Comment by tilgalas** [comment]
> @tilgalas Thank you for working on this! Would you mind to first remove the `"nullness"` value in the `@SuppressWarnings` annotation for the involved classes in this PR? Please consider this suggestion as non-blocking but I would like to see if this is possible with your helpful changes. As a second phase, may we also consider removing the `"rawtypes"`? Again, non-blocking but it's something we would like to strive for to improve the code quality.

sure, with pleasure!

**Comment by Abacn** [comment]
waiting on author

**Comment by tilgalas** [comment]
added a refactoring commit on top of the main one, which deals with most of the nullness and rawtypes warnings in the classes involved - I'm happy to move it to a PR of its own if the reviewers find the resulting changes too large to safely review

**Comment by tilgalas** [comment]
Run Java PreCommit

**Comment by tilgalas** [comment]
hi reviewers, I'm going to split this PR and put the nullness and rawness refactoring commit into its own PR

**Comment by tilgalas** [comment]
since there's a #32757, I'm closing this PR

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java [hunk 1]
```diff
@@ -836,10 +838,10 @@ public int nextFieldId() {
     }
 
     @Internal
-    public Row withFieldValueGetters(
-        Factory<List<FieldValueGetter>> fieldValueGetterFactory, Object getterTarget) {
+    public <T extends @NonNull Object> Row withFieldValueGetters(
```
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java:841]: unclear on this bound as well

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java [hunk 2]
```diff
@@ -42,45 +42,43 @@
...
+      Schema schema, Factory<List<FieldValueGetter<T, ?>>> getterFactory, T getterTarget) {
     super(schema);
     this.getterTarget = getterTarget;
     this.getters = getterFactory.create(TypeDescriptor.of(getterTarget.getClass()), schema);
   }
 
   @Override
   @SuppressWarnings({"TypeParameterUnusedInFormals", "unchecked"})
-  public <T> @Nullable T getValue(int fieldIdx) {
+  public @Nullable Object getValue(int fieldIdx) {
```
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java:59]: why remove type parameter?

File: sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java [hunk 3]
```diff
@@ -486,10 +518,17 @@ public static Schema toBeamSchema(org.apache.avro.Schema schema) {
     return builder.build();
   }
 
+  @EnsuresNonNullIf(
+      expression = {"#1"},
+      result = false)
+  private static boolean isNullOrEmpty(@Nullable String str) {
+    return str == null || str.isEmpty();
+  }
```
Comment (reuvenlax) [inline sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java:526]: what is wrong with Strings.isNullOrEmpty?

Comment (tilgalas) [reply sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java:526]: for some reason the checkerframework doesn't understand the nullness assertion when using it:
```
/usr/local/google/home/mszwaja/src/beam/sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java:538: error: [argument] incompatible argument for parameter namespace of AvroUtils.toAvroField.
      org.apache.avro.Schema.Field recordField = toAvroField(field, childNamespace);
                                                                    ^
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull String
/usr/local/google/home/mszwaja/src/beam/sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java:541: error: [argument] incompatible argument for parameter name of Schema.createRecord.
    return org.apache.avro.Schema.createRecord(schemaName, "", schemaNamespace, false, fields);
                                               ^
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull String
  ```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java (hunk 2 above)
Comment (tilgalas) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java:59]: I guess it's probably a matter of taste - my view is that this class can't possibly know what type of value a particular getter is going to return, so it's not commiting to anything by making the method generic - "I only know it's an Object, if you need a specific type, you'll have to cast it yourself" - I should probably also change the type of the getters list to `List<FieldValueGetter<T, Object>>` but then it would probably open another can of worms of covariant types, that is whether a, say, `FieldValueGetter<T, String>` is a `FieldValueGetter<T, Object>` - the current type of `List<FieldValueGetter<T, ?>>` is slightly incorrect as the `?` wildcard means an unknown type, but which is at least shared by all members of the list (which is not true)

I can understand the approach where the caller requests a specific type via a generic type parameter and is happy to accept the method throwing a ClassCastException if they had wrong assumptions about the getter - let me know if you want me to make it back a generic method

Comment (tilgalas) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java:59]: I feel like I will change the type of the list elements to `FieldValueGetter<T, Object>` anyway, as I find it more correct

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueGetter.java [hunk 4]
```diff
@@ -29,7 +30,7 @@
  * <p>Implementations of this interface are generated at runtime to map object fields to Row fields.
  */
 @Internal
-public interface FieldValueGetter<ObjectT, ValueT> extends Serializable {
+public interface FieldValueGetter<ObjectT extends @NonNull Object, ValueT> extends Serializable {
```
[the remark this answers is not in the document]
Comment (tilgalas) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueGetter.java:33]: so apparently there's a difference (I don't understand why it would be so, though) between a `@NonNull T` and `T extends @NonNull Object` - here's an example, consider:
```
  public interface I<T extends @NonNull Object> {
      T getT();
  }
```
with a signature as above, we can only write:
```
  public static <T extends @NonNull Object> void f(I<T> i) {
    System.out.println(i.getT().getClass());
  }
```
or alternatively:
```
  public static <T> void f(I<@NonNull T> i) {
    System.out.println(i.getT().getClass());
  }
``` 
The checkerframework will make sure that we propagate the non-nullness of the parameter. 
This won't work for example:
```
  public static <@Nullable T> void f(I<T> i) {
    T t = i.getT();
    if (t != null) {
      System.out.println(t.getClass());
    }
  }
```
nor won't this:
```
  public static <T> void f(I<@Nullable T> i) {
    T t = i.getT();
    if (t != null) {
      System.out.println(t.getClass());
    }
  }
```
but it will work if I change the signature of `I` to:
```
  public interface I<@NonNull T> {
      T getT();
  }
```
in other words, the `@NonNull T` annotation on the interface is not really enforced anywhere, it's more like a suggestion, and so it's not really expressing the concept of the interface only works with `@NonNull` types

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java (hunk 1 above)
Comment (tilgalas) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java:841]: I could change it to `<@NonNull T>` I suppose - it compiles, but then my IDE complains that "Non-null type argument is expected" on the `T` parametor of the `FieldValueGetter<T, ?>` type

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/GetterBasedSchemaProvider.java [hunk 5]
```diff
@@ -112,22 +111,28 @@ public SchemaUserTypeCreator schemaTypeCreator(
...
+                  (typeDescriptor, schema1) ->
+                      (List)
+                          GetterBasedSchemaProvider.this.fieldValueGetters(
+                              typeDescriptor, schema1));
     }
 
     @Override
     public Row apply(T input) {
-      return Row.withSchema(schema).withFieldValueGetters(getterFactory, input);
+      return Row.withSchema(schema).withFieldValueGetters((Factory) getterFactory, input);
```
[the remark this answers is not in the document]
Comment (tilgalas) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/GetterBasedSchemaProvider.java:135]: yeah, good question - I'm playing with that code right now and I think I could get rid of it actually

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/CachingFactory.java [hunk 6]
```diff
@@ -39,9 +41,9 @@
 public class CachingFactory<CreatedT> implements Factory<CreatedT> {
   private transient @Nullable ConcurrentHashMap<TypeDescriptor<?>, CreatedT> cache = null;
 
-  private final Factory<CreatedT> innerFactory;
+  private final @NotOnlyInitialized Factory<CreatedT> innerFactory;
```
[the remark this answers is not in the document]
Comment (tilgalas) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/CachingFactory.java:44]: so in `RowValueGettersFactory` there's this code:
```
    RowValueGettersFactory(Factory<List<FieldValueGetter<T, Object>>> gettersFactory) {
      this.gettersFactory = gettersFactory;
      this.cachingGettersFactory = new CachingFactory<>(this);
    }
```
since we're in the constructor, the `this` parameter is still not fully initialized, so by using this annotation we can temporarily (for the duration of the initalization of the CachingFactory object) allow assigning the `innerFactory` field a value which itself is under initialization (or indeed whose initialization status is unknown, as the signature of the CachingFactory constructor tells us)

see also https://checkerframework.org/manual/#circular-initialization

**Comment by ahmedabu98** [comment]
Fixes #32081

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueGetter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueGetter.java:33]: Generally in Beam I've mostly seen us using @Nullable, not @NonNull. I believe that NonNull is assumed unless @Nullable is specified. Are you not finding that to be the case?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/GetterBasedSchemaProvider.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/GetterBasedSchemaProvider.java:135]: ok

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java (hunk 2 above)
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java:59]: The base class uses a parameter here, so I think for now we should just match the base class.

File: sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java (hunk 3 above)
Comment (reuvenlax) [reply sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java:526]: I don't understand what this example has to do with isNullOrEmpty (which returns a boolean)?

Comment (tilgalas) [reply sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java:526]: this is the error that appears when I use Strings.isNullOrEmpty

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueGetter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tilgalas) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueGetter.java:33]: yeah, at least not when we're talking type parameters, for example:
```
  public interface I<T> {
      T getT();
  }

  public static <T> void f(I<T> i) {
    System.out.println(i.getT().getClass());
  }
  ```
  gives me 
  ```
error: [dereference.of.nullable] dereference of possibly-null reference i.getT()
    System.out.println(i.getT().getClass());
                             ^
  ```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java (hunk 1 above)
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java:841]: I think I prefer @NonNull T (is this simply a weakness in the IDE?), but I guess I don't have strong feelings here.

File: sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java (hunk 3 above)
Comment (reuvenlax) [reply sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java:526]: understood

**Comment by tilgalas** [comment]
Run Java PreCommit

**Comment by tilgalas** [comment]
Run Java PreCommit

**Comment by tilgalas** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by reuvenlax** [comment]
lgtm

File: sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java [hunk 7]
```diff
@@ -451,8 +484,7 @@ public static Field toBeamField(org.apache.avro.Schema.Field field) {
   public static org.apache.avro.Schema.Field toAvroField(Field field, String namespace) {
     org.apache.avro.Schema fieldSchema =
         getFieldSchema(field.getType(), field.getName(), namespace);
-    return new org.apache.avro.Schema.Field(
-        field.getName(), fieldSchema, field.getDescription(), (Object) null);
+    return new org.apache.avro.Schema.Field(field.getName(), fieldSchema, field.getDescription());
```
Comment (damccorm) [inline sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java:487]: This seems to break https://github.com/apache/beam/actions/workflows/beam_PostCommit_Java_Avro_Versions.yml

I think it is because in some older versions of avro, the constructor is different - https://avro.apache.org/docs/1.8.2/api/java/org/apache/avro/Schema.Field.html

**Comment by Abacn** [comment]
This likely breaks DataflowTemplate unit test: https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/2014/checks?check_run_id=33057418339

I'm trying to understand the scope of breaking change

**Comment by kennknowles** [comment]
This whole thread of work is interesting. FWIW the whole point of `TypeDescriptor` is to carry the generic information. Otherwise we could just use Java's reflective capabilities. Is there a TL;DR document about the limitation that you are improving?

**Comment by kennknowles** [comment]
> This whole thread of work is interesting. FWIW the whole point of `TypeDescriptor` is to carry the generic information. Otherwise we could just use Java's reflective capabilities. Is there a TL;DR document about the limitation that you are improving?

Ah, I see there is a doc linked from https://github.com/apache/beam/pull/31785

**Comment by kennknowles** [comment]
(also worth noting that `TypeDescriptor` is forked from Guava's TypeToken circa 2014 or 2015 in case that wasn't known or obvious)
